File: coins2.py
// Time Complexity : O(n*m)
// Space Complexity : O(n*m)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach: Make use of dp to find number of ways to make a given amount.



class Solution(object):
    def change(self, amount, coins):
        """
        :type amount: int
        :type coins: List[int]
        :rtype: int
        """
        
        dp=[[0 for i in range(amount+1)] for j in range(len(coins)+1)]
        
        for i in range(len(dp)):
            dp[i][0]=1
        
        for i in range(1,len(dp[0])):
            
            dp[0][i]=0
        
        
        for i in range(1,len(dp)):
            for j in range(1,len(dp[0])):
                if(coins[i-1]> j):
                    dp[i][j]= dp[i-1][j]
                else:
                    dp[i][j] = dp[i-1][j] + dp[i][j-coins[i-1]]
        return(dp[len(dp)-1][len(dp[0])-1])
        

File: coins2.java
// Time Complexity : O(mn) where m is the amount and n is the number of coins
// Space Complexity : O(mn) matrix
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : Pattern finding
/* Your code here along with comments explaining your approach: We do tabulation by number of coins as the rows and amount as columns.
The first column is filled with 1's because of calculation sake. If the coin is greater than the amount, then the value at the current cell
will be the same value in the above row else the value will be the value from previous row plus the value that is dp[j-coins[i]] steps back becuase
these are the number of ways that this coin can get j amount.
*/
class Solution {
    public int change(int amount, int[] coins) {
    int[][] dp = new int[coins.length+1][amount+1]; 
    for(int i = 0; i < dp.length;i++){
        dp[i][0] = 1;
    }
    for(int i = 1; i < dp.length; i++){
        for(int j =1; j < dp[0].length; j++){
            if(j < coins[i-1]){
            dp[i][j]=dp[i-1][j];                    // if amount > coins value
            } else {
                dp[i][j]= dp[i-1][j] + dp[i][j-coins[i-1]];         //number of ways coins[i] can make the amount j
            }
        }
    }
            return dp[dp.length-1][dp[0].length-1];                 // the last value gives the final answer
    }
}